tools/graphiteindex/src/main/java/com/criteo/biggraphite/graphiteindex/GraphiteSASIBuilder.java

```diff
@@ -0,0 +1,133 @@
+import org.apache.cassandra.db.RowIndexEntry;
+import org.apache.cassandra.db.compaction.CompactionInfo;
+import org.apache.cassandra.db.compaction.CompactionInterruptedException;
+import org.apache.cassandra.db.compaction.OperationType;
+import org.apache.cassandra.index.SecondaryIndexBuilder;
+import org.apache.cassandra.io.FSReadError;
+import org.apache.cassandra.io.sstable.KeyIterator;
+import org.apache.cassandra.io.sstable.SSTableIdentityIterator;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.io.util.RandomAccessReader;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.UUIDGen;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.apache.cassandra.io.util.FileHandle;
+
+public class GraphiteSASIBuilder
+    extends SecondaryIndexBuilder
+{
+    private static final Logger logger = LoggerFactory.getLogger(GraphiteSASIBuilder.class);
+
+    private final ColumnFamilyStore cfs;
+    private final ColumnDefinition column;
+    private final SortedSet<SSTableReader> sstables;
+
+    private final UUID compactionId;
+    private final long totalBytes;
+    private long processedBytes;
+
+    public GraphiteSASIBuilder(
+        ColumnFamilyStore cfs, ColumnDefinition column, SortedSet<SSTableReader> sstables
+    )
+    {
+        long totalBytes = 0;
+        for (SSTableReader reader : sstables) {
+            totalBytes += reader.getIndexFile().dataLength();
+        }
+
+        this.cfs = cfs;
+        this.column = column;
+        this.sstables = sstables;
+
+        this.compactionId = UUIDGen.getTimeUUID();
+        this.totalBytes = totalBytes;
+        this.processedBytes = 0;
+    }
+
+    @Override public CompactionInfo getCompactionInfo()
+    {
+        return new CompactionInfo(
+            cfs.metadata,
+            OperationType.INDEX_BUILD,
+            processedBytes,
+            totalBytes,
+            compactionId
+        );
+    }
+
+    @Override public void build()
+    {
+        sstables.forEach(this::buildIndex);
+    }
+
+    protected void buildIndex(SSTableReader sstable)
+    {
+        try (
+            RandomAccessReader dataFile = sstable.openDataReader();
+            KeyIterator keys = new KeyIterator(sstable.descriptor, cfs.metadata);
+        ) {
+            String indexName = GraphiteSASI.makeIndexName(column, sstable.descriptor.generation);
+            logger.debug("Building index {}", indexName);
+
+            Path indexPath = new File(sstable.descriptor.directory, indexName).toPath();
+            MetricsIndex index = new MetricsIndex(indexName, Optional.of(indexPath));
+
+            PerSSTableIndexWriter indexWriter = new PerSSTableIndexWriter(
+                sstable.descriptor, OperationType.COMPACTION, column, index
+            );
+
+            long previousKeyPosition = 0;
+            while (keys.hasNext()) {
+                if (isStopRequested()) {
```

Comment:
maybe put the content of the while in a separate function ?

Reply:
kinda done